bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/PageImpl.java

```diff
@@ -0,0 +1,446 @@
+import com.adobe.cq.wcm.core.components.models.Page;
+import com.day.cq.wcm.api.Template;
+import com.day.cq.wcm.api.TemplatedResource;
+import com.day.cq.wcm.api.components.ComponentContext;
+import com.day.cq.wcm.api.designer.Style;
+import com.day.cq.wcm.api.policies.ContentPolicy;
+import com.day.cq.wcm.api.policies.ContentPolicyManager;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = {Page.class, ContainerExporter.class}, resourceType = PageImpl.RESOURCE_TYPE)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME, extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class PageImpl extends com.adobe.cq.wcm.core.components.internal.models.v2.PageImpl implements Page {
+
+    protected static final String RESOURCE_TYPE = "core/wcm/components/page/v3/page";
+
+    /**
+     * Flags the child pages. Optionally available as a request attribute
+     */
+    private static final String ATTR_IS_CHILD_PAGE = "com.adobe.cq.wcm.core.components.internal.models.HierarchyPage.isChildPage";
+
+    /**
+     * Is the current model to be considered as a model root
+     */
+    private static final String PR_IS_ROOT = "isRoot";
+
+    /**
+     * URL extension specific to the Sling Model exporter
+     */
+    private static final String URL_MODEL_EXTENSION = ".model.json";
+
+    /**
+     * Request attribute key of the request page entry point
+     */
+    private static final String ATTR_HIERARCHY_ENTRY_POINT_PAGE = "com.adobe.cq.wcm.core.components.internal.models.HierarchyPage.entryPointPage";
+
+    @ScriptVariable
+    private Resource resource;
+
+    private Map<String, ? extends Page> childPages = null;
+
+    @Nullable
+    @Override
+    public String getExportedHierarchyType() {
+        return HierarchyTypes.PAGE;
+    }
+
+    @Nonnull
+    @Override
+    public Map<String, ? extends Page> getExportedChildren() {
+        if (childPages == null) {
+            childPages = getChildPageModels(request);
+        }
+
+        return childPages;
+    }
+
+    @Nonnull
+    @Override
+    public String getExportedPath() {
+        return currentPage.getPath();
+    }
+
+    @Nullable
+    @Override
+    public String getRootUrl() {
+        if (currentStyle != null && currentStyle.containsKey(PR_IS_ROOT)) {
+            return getModelUrl(request, currentPage);
+        }
+
+        com.day.cq.wcm.api.Page page = getRootPage();
+
+        if (page != null) {
+            return getModelUrl(request, page);
+        }
+
+        return null;
+    }
+
+    /**
+     * @return Returns the root model of the given page
+     */
+    @Nullable
+    @Override
+    public Page getRootModel() {
+        if (currentStyle != null && currentStyle.containsKey(PR_IS_ROOT)) {
+            return this;
+        }
+
+        com.day.cq.wcm.api.Page rootPage = getRootPage();
+
+        if (rootPage == null) {
+            return null;
+        }
+
+        PageImpl.requestSetHierarchyEntryPoint(request, currentPage);
```

Comment:
It seems counter-intuitive for a getter to be changing the state of the request. You could pass the currentPage (entry point in the hierarchy) to the request wrapper on next lines. You could also reset any custom attributes after the logic executes.